WordPress/gutenberg
# Experiment: Auto-inserting block patterns on the frontend

## What?
Spun off from #50103, this PR explores auto-inserting block _patterns_ (rather than just blocks).

## Why?
For the rationale of why block patterns might be better suited for auto-insertion, see https://github.com/WordPress/gutenberg/pull/50103#issuecomment-1579060416. For general auto-inserting concepts, see https://github.com/WordPress/gutenberg/issues/39439.

## How?
Much like #50103, we use:
- the `render_block` hook to auto-insert a block pattern before or after a given "anchor" block, and
- the  `render_block_data` hook (which -- unlike the `render_block` hook -- conveniently gives us access to a given block's children) to auto-insert a block pattern as an anchor block's first or last child.

In both filters, we iterate over all registered block patterns to find any that specify the current block as their "anchor" block. 

To illustrate both concepts, we add:
- one block pattern featuring the Social Icon block that we insert after the Post Content block, and
- one block pattern featuring the Avatar block that we insert as the Comment Template block's last child.

The latter example is chosen to also demonstrate that block context is successfully passed to auto-inserted blocks (as evidenced by the auto-inserted Avatar blocks showing each comment author's avatar).

## Testing Instructions

- Make sure to rebuild blocks (`npm run build`).
- Use the Twenty Twenty-Three theme.
- On a single post page, you should see the Social Icon block (showing the WordPress logo) below the post (as sort of a stand-in for a Like button 😅)
- Have a look at comments below a given post. You should see an Avatar block showing the comment author's avatar (as a stand-in for a Comment Like button)

![image](https://github.com/WordPress/gutenberg/assets/96308/6629dcba-a62a-41f4-beef-e275ffca498c)
